File: comfy_api_nodes/shot_presets.py
"""Cinematic shot-preset catalog + prompt compiler for FilmShotNode.

Mirrors frontend/app/data/shot-presets.ts (gallery tiles). Keep `id` identical
to the TS catalog — it's the dispatch key the gallery writes into the node's
`preset` widget.

Each preset is a complete framing recipe across five dimensions (size, angle,
movement, lens, composition) plus a mood note. `build_shot_phrase` compiles a
recipe into prompt language; the dialect arg picks per-model phrasing
(standard prose / Veo lens-forward / Hailuo bracket commands).

Design doc: docs/plans/2026-06-10-film-a-shot-node-design.md
"""
from __future__ import annotations

from dataclasses import dataclass


# Sentinel for the ADVANCED override combos: keep the preset's own value.
AUTO = "auto (preset)"

DEFAULT_PRESET_ID = "push-in"


@dataclass(frozen=True)
class ShotPreset:
    id: str
    label: str
    category: str      # movement | angle | lens | composition
    size: str          # "medium close-up"
    angle: str         # composes as f"from {angle}" — write accordingly
    movement: str      # full clause: "the camera slowly dollies in"
    lens: str          # "50mm lens with shallow depth of field"
    composition: str   # "subject centered in frame"
    note: str          # mood line: "builds quiet tension"


PRESETS: list[ShotPreset] = [
    # ----- Movement-led ------------------------------------------------------
    ShotPreset("push-in", "Slow push-in", "movement",
        "medium close-up", "eye level",
        "the camera slowly dollies in toward the subject",
        "50mm lens with shallow depth of field",
        "subject centered in frame", "builds quiet tension"),
    ShotPreset("pull-back", "Pull-back reveal", "movement",
        "close-up widening to a wide shot", "eye level",
        "the camera steadily dollies out, revealing the surroundings",
        "35mm lens with deepening focus",
        "subject anchored in place as the world grows around them",
        "the context lands at the end"),
    ShotPreset("crane-reveal", "Crane reveal", "movement",
        "wide shot", "a low position rising high",
        "the camera cranes up smoothly",
        "24mm lens with deep focus",
        "landscape framing with a strong horizon", "establishing grandeur"),
    ShotPreset("orbit", "Hero orbit", "movement",
        "medium shot", "eye level",
        "the camera arcs in a slow 180-degree orbit around the subject",
        "35mm lens with shallow depth of field",
        "subject locked at frame center", "the hero moment"),
    ShotPreset("tracking", "Lateral tracking", "movement",
        "medium shot in profile", "eye level",
        "the camera tracks laterally alongside the moving subject, matched to their pace",
        "40mm lens with deep focus",
        "leading room ahead of the subject", "walk-and-talk energy"),
    ShotPreset("handheld", "Handheld urgency", "movement",
        "medium close-up", "eye level",
        "shaky handheld camera following the subject",
        "28mm lens",
        "loose, imperfect framing", "documentary urgency"),
    ShotPreset("dolly-zoom", "Dolly zoom (Vertigo)", "movement",
        "medium close-up", "eye level",
        "the camera dollies in while the lens zooms out",
        "50mm lens, the background perspective visibly warping",
        "subject locked dead center", "reality bends around them"),
    ShotPreset("tilt-reveal", "Tilt-up reveal", "movement",
        "full shot revealed from the feet upward", "a low angle",
        "the camera tilts up slowly from the ground to the face",
        "35mm lens",
        "vertical reveal framing", "sizing them up"),
    ShotPreset("whip-pan", "Whip pan", "movement",
        "medium shot", "eye level",
        "the camera whips violently sideways in a fast pan, streaking with motion blur",
        "35mm lens",
        "framing snaps from one point to the next", "an energy spike"),
    ShotPreset("crash-zoom", "Crash zoom", "movement",
        "wide shot punching in to a close-up", "eye level",
        "an abrupt crash zoom punches toward the subject",
        "zoom lens",
        "subject suddenly fills the frame", "a grindhouse exclamation mark"),
    ShotPreset("snorricam", "Snorricam", "movement",
        "close-up, body-rigged", "a body-rigged mount facing the actor",
        "a snorricam locked to the actor's body, the world lurching and swimming behind them",
        "28mm lens",
        "face pinned center while the background reels", "panic and unraveling"),
    ShotPreset("steadicam-oner", "Steadicam oner", "movement",
        "medium shot following the subject", "eye level",
        "a flowing steadicam glide that follows unbroken through doorways and spaces",
        "32mm lens with deep focus",
        "continuously reframing around the moving subject", "the long-take feel"),
    ShotPreset("fpv-dive", "FPV drone dive", "movement",
        "wide shot tightening rapidly", "a plunging aerial angle",
        "an FPV drone dives aggressively and weaves toward the subject",
        "ultra-wide lens",
        "horizon rolling with the dive", "pure adrenaline"),
    ShotPreset("aerial-orbit", "Aerial establish orbit", "movement",
        "extreme wide shot", "a high aerial angle",
        "a drone circles the location in a slow orbit",
        "24mm lens with deep focus",
        "the landscape framed like a map coming alive", "the opening-credits shot"),
    ShotPreset("ground-rush", "Ground-rush tracking", "movement",
        "low medium shot", "inches off the ground",
        "the camera skims fast and low across the surface toward the subject",
        "24mm lens",
        "ground rushing through the lower frame", "road-blur menace"),

    # ----- Angle-led ---------------------------------------------------------
    ShotPreset("god-shot", "Overhead god shot", "angle",
        "wide shot", "directly overhead, bird's-eye",
        "the camera descends slowly straight down",
        "24mm lens with deep focus",
        "geometric framing of the ground below", "fate watching from above"),
    ShotPreset("low-hero", "Low-angle power", "angle",
        "medium shot", "a strong low angle looking up",
        "the camera pushes in slightly",
        "24mm wide-angle lens with mild distortion",
        "the subject towering over the frame", "an imposing entrance"),
    ShotPreset("dutch", "Dutch drift", "angle",
        "medium close-up", "a dutch tilt of about 15 degrees",
        "the camera drifts slowly sideways",
        "40mm lens with shallow depth of field",
        "off-balance rule-of-thirds framing", "something is quietly wrong"),
    ShotPreset("worms-eye", "Worm's-eye sky", "angle",
        "extreme wide shot", "looking straight up from the ground",
        "the camera holds static, rolling slowly",
        "18mm ultra-wide lens",
        "towers and sky swallowing the frame", "vertigo in reverse"),

    # ----- Lens-led ----------------------------------------------------------
    ShotPreset("anamorphic", "Anamorphic dream", "lens",
        "medium shot", "eye level",
        "the camera drifts slowly",
        "anamorphic lens with horizontal blue flares and oval bokeh",
        "2.39:1 widescreen letterbox framing", "prestige-film sheen"),
    ShotPreset("macro", "Macro detail", "lens",
        "extreme close-up", "eye level",
        "a focus pull racks across the detail",
        "macro lens with razor-thin depth of field",
        "the single detail isolated against soft blur", "the object tells the story"),
    ShotPreset("rack-focus", "Rack focus reveal", "lens",
        "medium close-up across two depth planes", "eye level",
        "the frame holds still while focus pulls from foreground to background",
        "85mm lens with razor-thin depth of field",
        "two subjects stacked in depth", "attention is the edit"),
    ShotPreset("telephoto", "Telephoto compression", "lens",
        "medium close-up from far away", "eye level",
        "the camera holds nearly still with a slight pan",
        "300mm telephoto lens compressing the planes",
        "blurred foreground passers-by stacking the depth", "surveillance distance"),

    # ----- Composition-led ---------------------------------------------------
    ShotPreset("locked-off", "Symmetrical one-point", "composition",
        "wide shot", "eye level",
        "a locked-off static camera, perfectly still",
        "32mm lens with deep focus",
        "dead-center one-point-perspective symmetry", "an unblinking formal stare"),
    ShotPreset("ots", "Over-the-shoulder", "composition",
        "medium close-up", "eye level",
        "the camera holds nearly still with a micro drift",
        "65mm lens with shallow depth of field",
        "framed over a foreground shoulder", "conversation intimacy"),
    ShotPreset("pov", "POV walk", "composition",
        "first-person point of view", "eye level",
        "handheld camera walking forward as the character's own eyes",
        "28mm lens",
        "hands and body edges intruding at the frame borders", "you are there"),
    ShotPreset("voyeur-frame", "Voyeur doorframe", "composition",
        "medium shot", "eye level",
        "a static camera watching through a doorway",
        "50mm lens",
        "framed through a door or window slit, dark edges crowding the subject",
        "being watched"),
    ShotPreset("mirror", "Mirror double", "composition",
        "medium close-up", "eye level",
        "the camera holds still, pushing in slowly",
        "50mm lens",
        "the subject and their mirror reflection sharing the frame",
        "two truths at once"),
]

PRESETS_BY_ID: dict[str, ShotPreset] = {p.id: p for p in PRESETS}
PRESET_IDS: list[str] = [p.id for p in PRESETS]


# ---------- ADVANCED override option lists ----------------------------------
# Each combo's options ARE the substitution phrases — picking one replaces that
# dimension of the active preset verbatim. AUTO keeps the preset's value.

SIZE_OPTIONS = [AUTO, "extreme close-up", "close-up", "medium close-up",
                "medium shot", "full shot", "wide shot", "extreme wide shot"]

ANGLE_OPTIONS = [AUTO, "eye level", "a low angle", "a high angle",
                 "directly overhead, bird's-eye", "a dutch tilt", "ground level"]

MOVEMENT_OPTIONS = [AUTO,
    "a locked-off static camera",
    "the camera slowly dollies in toward the subject",
    "the camera steadily dollies out",
    "the camera pans slowly left to right",
    "the camera tilts up slowly",
    "the camera cranes up smoothly",
    "the camera arcs in a slow orbit around the subject",
    "the camera tracks laterally alongside the subject",
    "shaky handheld camera following the subject",
    "a flowing steadicam glide",
    "an FPV drone dives toward the subject",
    "an abrupt crash zoom punches toward the subject"]

LENS_OPTIONS = [AUTO, "18mm ultra-wide lens", "24mm wide-angle lens", "35mm lens",
                "50mm lens with shallow depth of field",
                "85mm lens with razor-thin depth of field",
                "300mm telephoto lens compressing the planes",
                "anamorphic lens with horizontal blue flares and oval bokeh",
                "macro lens with razor-thin depth of field"]

COMPOSITION_OPTIONS = [AUTO, "subject centered in frame", "rule-of-thirds framing",
                       "dead-center one-point-perspective symmetry",
                       "framed over a foreground shoulder",
                       "first-person POV framing",
                       "framed through a doorway",
                       "leading room ahead of the subject"]

File: frontend/app/data/shot-presets.ts
/**
 * Shot-preset catalog — drives the "Film a shot" node's gallery. Each preset
 * is a complete cinematography recipe (size · angle · movement · lens ·
 * composition); the backend compiles it into model-dialect prompt language.
 *
 * Mirrors comfy_api_nodes/shot_presets.py — keep `id` identical (dispatch key).
 * Adding a preset: append an entry in BOTH files.
 *
 * `thumb` drives the data-only CSS thumbnail in the gallery card (person
 * silhouette + motion arrow + optional overlay). No image assets.
 */

export type ShotCategory = 'movement' | 'angle' | 'lens' | 'composition'

export const SHOT_CATEGORY_LABELS: Record<ShotCategory, string> = {
  movement: 'Movement',
  angle: 'Angle',
  lens: 'Lens',
  composition: 'Composition',
}

export type ShotArrow =
  | 'in' | 'out' | 'up' | 'upright' | 'down' | 'right'
  | 'orbit' | 'shake' | 'dive' | 'flow' | 'rack' | 'none'

export type ShotOverlay =
  | 'doorframe' | 'mirror' | 'blur' | 'streak' | 'shoulder' | 'hands' | 'none'

export interface ShotThumb {
  scale?: number    // person scale (1 = medium); 0 hides the person
  top?: number      // person top offset in % (default 22)
  tilt?: number     // frame tilt in degrees
  arrow?: ShotArrow
  overlay?: ShotOverlay
}

export interface ShotPreset {
  id: string
  label: string
  category: ShotCategory
  recipe: string    // one-line recipe summary shown on the card
  pitch: string     // mood line
  thumb: ShotThumb
}

export const SHOT_PRESETS: ShotPreset[] = [
  // ── Movement ──────────────────────────────────────────────────────────────
  { id: 'push-in', label: 'Slow push-in', category: 'movement',
    recipe: 'MCU · eye level · slow dolly in · 50mm shallow · centered',
    pitch: 'Builds quiet tension', thumb: { scale: 1.3, arrow: 'in' } },
  { id: 'pull-back', label: 'Pull-back reveal', category: 'movement',
    recipe: 'CU → wide · dolly out · 35mm deepening · subject anchored',
    pitch: 'The context lands at the end', thumb: { scale: 0.7, arrow: 'out' } },
  { id: 'crane-reveal', label: 'Crane reveal', category: 'movement',
    recipe: 'Wide · rising low → high · crane up · 24mm deep',
    pitch: 'Establishing grandeur', thumb: { scale: 0.55, top: 40, arrow: 'upright' } },
  { id: 'orbit', label: 'Hero orbit', category: 'movement',
    recipe: 'MS · slow 180° arc · 35mm shallow · subject locked center',
    pitch: 'The hero moment', thumb: { scale: 1, arrow: 'orbit' } },
  { id: 'tracking', label: 'Lateral tracking', category: 'movement',
    recipe: 'MS profile · side-track with subject · 40mm deep · leading room',
    pitch: 'Walk-and-talk energy', thumb: { scale: 1, arrow: 'right' } },
  { id: 'handheld', label: 'Handheld urgency', category: 'movement',
    recipe: 'MCU · shaky handheld follow · 28mm · loose framing',
    pitch: 'Documentary urgency', thumb: { scale: 1.1, tilt: 1.5, arrow: 'shake' } },
  { id: 'dolly-zoom', label: 'Dolly zoom (Vertigo)', category: 'movement',
    recipe: 'MCU · dolly in + zoom out · warping background · centered',
    pitch: 'Reality bends around them', thumb: { scale: 1.2, arrow: 'in', overlay: 'streak' } },
  { id: 'tilt-reveal', label: 'Tilt-up reveal', category: 'movement',
    recipe: 'Feet → face · low angle · slow tilt up · 35mm',
    pitch: 'Sizing them up', thumb: { scale: 1.2, top: 8, arrow: 'up' } },
  { id: 'whip-pan', label: 'Whip pan', category: 'movement',
    recipe: 'MS · violent fast pan, motion-blur streaks · 35mm',
    pitch: 'An energy spike', thumb: { scale: 1, arrow: 'right', overlay: 'streak' } },
  { id: 'crash-zoom', label: 'Crash zoom', category: 'movement',
    recipe: 'Wide → CU · abrupt punch-in zoom · grindhouse',
    pitch: 'An exclamation mark', thumb: { scale: 1.5, top: 12, arrow: 'in' } },
  { id: 'snorricam', label: 'Snorricam', category: 'movement',
    recipe: 'CU body-rigged · face locked, world lurches · 28mm',
    pitch: 'Panic and unraveling', thumb: { scale: 1.4, top: 14, tilt: -4, arrow: 'shake' } },
  { id: 'steadicam-oner', label: 'Steadicam oner', category: 'movement',
    recipe: 'MS following · unbroken glide through spaces · 32mm deep',
    pitch: 'The long-take feel', thumb: { scale: 1, arrow: 'flow' } },
  { id: 'fpv-dive', label: 'FPV drone dive', category: 'movement',
    recipe: 'Wide → tight · plunging aerial dive + weave · ultra-wide',
    pitch: 'Pure adrenaline', thumb: { scale: 0.6, top: 46, arrow: 'dive' } },
  { id: 'aerial-orbit', label: 'Aerial establish orbit', category: 'movement',
    recipe: 'Extreme wide · high aerial · slow drone circle · 24mm deep',
    pitch: 'The opening-credits shot', thumb: { scale: 0.45, top: 48, arrow: 'orbit' } },
  { id: 'ground-rush', label: 'Ground-rush tracking', category: 'movement',
    recipe: 'Low MS · inches off the floor · fast forward skim · 24mm',
    pitch: 'Road-blur menace', thumb: { scale: 0.85, top: 16, arrow: 'in', overlay: 'streak' } },

  // ── Angle ────────────────────────────────────────────────────────────────
  { id: 'god-shot', label: 'Overhead god shot', category: 'angle',
    recipe: 'Wide · directly overhead · slow descend · geometric floor',
    pitch: 'Fate watching from above', thumb: { scale: 0.55, top: 42, arrow: 'down' } },
  { id: 'low-hero', label: 'Low-angle power', category: 'angle',
    recipe: 'MS · strong low angle · slight push · 24mm distortion',
    pitch: 'An imposing entrance', thumb: { scale: 1.35, top: 6, arrow: 'in' } },
  { id: 'dutch', label: 'Dutch drift', category: 'angle',
    recipe: 'MCU · dutch 15° · slow lateral drift · 40mm shallow',
    pitch: 'Something is quietly wrong', thumb: { scale: 1.1, tilt: -8, arrow: 'right' } },
  { id: 'worms-eye', label: "Worm's-eye sky", category: 'angle',
    recipe: 'Extreme low, looking straight up · 18mm · towers swallow the frame',
    pitch: 'Vertigo in reverse', thumb: { scale: 1.25, top: 2, arrow: 'none' } }, // camera rolls; no roll arrow available

  // ── Lens ─────────────────────────────────────────────────────────────────
  { id: 'anamorphic', label: 'Anamorphic dream', category: 'lens',
    recipe: 'MS · slow drift · anamorphic flares, oval bokeh · letterbox',
    pitch: 'Prestige-film sheen', thumb: { scale: 1, overlay: 'streak', arrow: 'none' } },
  { id: 'macro', label: 'Macro detail', category: 'lens',
    recipe: 'ECU · rack focus pull · macro shallow · isolated detail',
    pitch: 'The object tells the story', thumb: { scale: 0, overlay: 'blur', arrow: 'rack' } },
  { id: 'rack-focus', label: 'Rack focus reveal', category: 'lens',
    recipe: 'Two depth planes · static frame, focus pulls front → back · 85mm',
    pitch: 'Attention is the edit', thumb: { scale: 0.9, overlay: 'blur', arrow: 'rack' } },
  { id: 'telephoto', label: 'Telephoto compression', category: 'lens',
    recipe: 'MCU from afar · 300mm stacked planes · blurred passers-by',
    pitch: 'Surveillance distance', thumb: { scale: 0.95, overlay: 'blur', arrow: 'none' } },

  // ── Composition ──────────────────────────────────────────────────────────
  { id: 'locked-off', label: 'Symmetrical one-point', category: 'composition',
    recipe: 'Wide · locked-off static · 32mm deep · dead-center symmetry',
    pitch: 'An unblinking formal stare', thumb: { scale: 0.9, overlay: 'none', arrow: 'none' } }, // dead-center symmetry; thirds grid would contradict
  { id: 'ots', label: 'Over-the-shoulder', category: 'composition',
    recipe: 'MCU · 65mm shallow · framed over a foreground shoulder',
    pitch: 'Conversation intimacy', thumb: { scale: 1, overlay: 'shoulder', arrow: 'none' } },
  { id: 'pov', label: 'POV walk', category: 'composition',
    recipe: 'First person · handheld forward · 28mm · body edges in frame',
    pitch: 'You are there', thumb: { scale: 0, overlay: 'hands', arrow: 'right' } },
  { id: 'voyeur-frame', label: 'Voyeur doorframe', category: 'composition',
    recipe: 'MS · static through doorway slit · 50mm · dark edges crowd in',
    pitch: 'Being watched', thumb: { scale: 0.95, overlay: 'doorframe', arrow: 'none' } },
  { id: 'mirror', label: 'Mirror double', category: 'composition',
    recipe: 'MCU · subject + reflection share frame · 50mm · slow push',
    pitch: 'Two truths at once', thumb: { scale: 0.9, overlay: 'mirror', arrow: 'in' } },
]

export const SHOT_PRESETS_BY_ID: Record<string, ShotPreset> =
  Object.fromEntries(SHOT_PRESETS.map(p => [p.id, p]))

// Fallback for unknown/old workflow preset ids — mirrors DEFAULT_PRESET_ID in
// comfy_api_nodes/shot_presets.py.
export const DEFAULT_SHOT_PRESET_ID = 'push-in'
